Library/Homebrew/linkage_checker.rb

```diff
@@ -168,21 +168,29 @@ def check_dylibs(rebuild_cache:)
     keg_files_dylibs.each do |file, dylibs|
       file_has_any_rpath_dylibs = T.let(false, T::Boolean)
       dylibs.each do |dylib|
+        pathname = Pathname(file)
+        # Resolve some variable install names.
+        expanded_path = pathname.dirname
+        dylib.sub!("@loader_path", expanded_path.to_s)
+        dylib.sub!("@executable_path", expanded_path.to_s) if pathname.binary_executable?
+        # Remove redundant dots from the path.
+        dylib = Pathname(dylib).cleanpath.to_s
```

Comment:
I believe moving it further up the call stack to `#dynamically_linked_libraries` would make it so that the call to `#each_linkage_for` in `#fix_dynamic_linkage` will receive the resolved paths for any relative linkages and attempt to needlessly fix it again (using an inexistent name in the process)?

Reply:
We could give `#dynamically_linked_libraries` an optional kwarg, e.g. `resolve_variable_references`, so that calling
```
file.dynamically_linked_libraries
```
will return resolved paths, while
```
file.dynamically_linked_libraries(resolve_variable_references: false)
```
will not. You can use the latter in `#fix_dynamic_linkage`.

Reply:
Not opposed to it. I like the idea of moving this logic to the `MachOShim` class and controlling it via a `resolve_variable_references` kwarg (that defaults to `true`), although I think a name like that could mean the method may be expected to also resolve `@rpath`-based install names (which it currently won't and isn't needed to check the changes in this PR; although that could be added in a future improvement to `LinkageChecker`).

If you're okay with that (I'm fine with it), I can make this change.